core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java

```diff
@@ -2,52 +2,107 @@
 
 import static java.lang.String.format;
 
+import uk.gov.justice.services.core.handler.exception.HandlerExecutionException;
 import uk.gov.justice.services.core.handler.registry.exception.InvalidHandlerException;
 import uk.gov.justice.services.messaging.Envelope;
 
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 
 /**
- * Base class for synchronous and asynchronous handler methods.
+ * Encapsulates a handler class instance and a handler method.
+ *
+ * Asynchronous handler methods will return a null {@link Void} whereas synchronous handler methods
+ * must return an {@link Envelope}.
  */
-public abstract class HandlerMethod {
+public class HandlerMethod {
 
-    protected final Object handlerInstance;
-    protected final Method handlerMethod;
+    private final Object handlerInstance;
+    private final Method handlerMethod;
+
+    private final boolean isSynchronous;
 
     /**
-     * Constructor with common handler method validation.
+     * Constructor with handler method validation.
      *
-     * @param handlerInstance the instance of the handler object
-     * @param handlerMethod   the method on the handler object
+     * @param object             the instance of the handler object
+     * @param method             the method on the handler object
+     * @param expectedReturnType the expected return type for the method
      */
-    HandlerMethod(final Object handlerInstance, final Method handlerMethod) {
+    public HandlerMethod(final Object object, final Method method, final Class<?> expectedReturnType) {
 
-        if (handlerInstance == null) {
+        if (object == null) {
             throw new IllegalArgumentException("Handler instance cannot be null");
         }
 
-        if (handlerMethod == null) {
+        if (method == null) {
             throw new IllegalArgumentException("Handler method cannot be null");
         }
 
-        if (handlerMethod.getParameterTypes().length != 1) {
-            throw new InvalidHandlerException("Handles method must have exactly one parameter. Found " + handlerMethod.getParameterTypes().length);
+        if (method.getParameterTypes().length != 1) {
+            throw new InvalidHandlerException(
+                    format("Handles method must have exactly one parameter; found %d",
+                            method.getParameterTypes().length));
+        }
+
+        final Class<?> paramClass = method.getParameterTypes()[0];
+        if (paramClass != Envelope.class) {
+            throw new IllegalArgumentException("Handler methods must receive Envelope as an argument");
+        }
+
+        if (isVoid(expectedReturnType) && !isVoid(method.getReturnType())) {
```

Comment:
Error handling still needs to go somewhere! I'd prefer to keep it in one place. Going to try and simplify.